rfcs/2020-04-20-Optimizer-minimize.md

```diff
@@ -0,0 +1,293 @@
+      self,
+      loss,
+      variables,
+      tape=None,
+      all_reduce_sum_gradients=False):
+    if is_tensor(loss) and not tape:
+      raise ValueError('Must provide tape with tensor loss.')
+    tape = tape or GradientTape()
+    with tape:
+      if callable(loss):
+        loss = loss()
+      loss = self._transform_loss(loss) # A no-op in our built-in optimizers
+    gradients = self._get_gradients(loss, variables, tape)
+    grads_and_vars = zip(gradients, variables)
+    grads_and_vars = self._transform_unaggregated_gradients(grads_and_vars)
+    if all_reduce_sum_gradients:
+      grads_and_vars = self._aggregate_gradients(grads_and_vars)
+      grads_and_vars = self._transform_gradients(grads_and_vars)
+    return grads_and_vars
+
+  def apply_gradients(self, grads_and_vars, aggregate=True):
+    if aggregate:
+      grads_and_vars = self._aggregate_gradients(grads_and_vars)
+      grads_and_vars = self._transform_gradients(grads_and_vars)  # No-op by default
+    # By passing all_reduce_sum_gradients, only the Variable updates are run.
+    # This gives users complete control, in the case that they don't want to use
+    # the hooks provided.
+    self.apply_updates(grads_and_vars)
+```
+
+
+Use of Optimizer.minimize in Model.train_step:
+
+```python
+class Model:
+
+  def train_step(self, data):
+    data = expand_1d(data)
+    x, y, sample_weight = unpack_x_y_sample_weight(data)
+    with tf.GradientTape() as tape:
+       y_pred = self(x, training=True)
+       loss = self.compiled_loss(y, y_pred, sample_weight, self.losses)
+   self.optimizer.minimize(loss, self.trainable_variables, tape=tape)
+   self.compiled_metrics.update_state(y, y_pred, sample_weight)
+   return {m.name: m.result() for m in self.metrics}
+```
+
+Details of proposal:
+
+* Adds the ability to accept a loss Tensor and a GradientTape to Optimizer.minimize.
+
+* Maintains full backwards compatibility. When a callable loss is passed, simply create a GradientTape and call the loss inside it like currently done.
+
+* Add public Optimizer methods that can be overridden to support custom functionality for the steps outlined in the Background section:
+
+
+(1) `Optimizer._transform_loss`
+
+(2) `Optimizer._get_gradients`
+
+(3) `Optimizer._transform_unaggregated_gradients`
+
+(4) `Optimizer._aggregate_gradients`
+
+(5) `Optimizer._transform_gradients` (aggregated gradients)
+
+(6) `Optimizer._apply_updates` (calls existing existing _resource_apply_{dense|sparse})
+
+(a) Item (6) mirrors `Sonnet`’s apply method (i.e. is “just the math”)
+
+* Use Optimizer.minimize API in Model.fit
+
+* Optimizer.apply_gradients method is kept. For users who want to control all loss and gradient manipulation, and want the Optimizer to simply apply the Variable updates, they can call `Optimizer.apply_gradients(..., aggregate=False)`
+
+
+## Examples
+
+(1) Custom gradient clipping
+
+```python
+optimizer = tf.keras.optimizers.Adam(0.1, transform_gradients=my_gradient_clipping)
```

Comment:
But what goes into my_gradient_clipping? Examples would be helpful from the user perspective, so that we don't tell people, "Oh, you just pass a thing," and have them find out later, "The thing doesn't match my needs."

Reply:
Done, fleshed out example